ttcn3/types/types.go

```diff
@@ -179,6 +179,35 @@ type ListType struct {
 }
 
 func (t *ListType) String() string {
+	switch t.Kind {
+	case RecordOf:
+		var lengthConstraint string = " "
+		if t.LengthConstraint.Expr != nil {
+			lengthConstraint = " length(" + t.LengthConstraint.String() + ") "
+		}
+		return "record" + lengthConstraint + "of " + t.ElementType.String()
+	case SetOf:
+		var lengthConstraint string = " "
+		if t.LengthConstraint.Expr != nil {
+			lengthConstraint = " length(" + t.LengthConstraint.String() + ") "
+		}
+		return "set" + lengthConstraint + "of " + t.ElementType.String()
+	case Charstring, Octetstring, Hexstring, Bitstring, UniversalCharstring:
+		var lengthConstraint string = ""
+		if t.LengthConstraint.Expr != nil {
+			lengthConstraint = " length(" + t.LengthConstraint.String() + ")"
+		}
+		return t.Kind.String() + lengthConstraint
```

Comment:
You probably can merge those alternatives into a single path for Kind != Array:

```suggestion
	default:
		var lengthConstraint string = " "
		if t.LengthConstraint.Expr != nil {
			lengthConstraint = " length(" + t.LengthConstraint.String() + ")"
		}
		return t.Kind.String() + lengthConstraint + " of " + t.ElementType.String()
```